### PR TITLE
feat: add model-based algorithms

### PR DESCRIPTION
**Pull request by hdadong**
feat: add model-based algorithms

## Types of changes

What types of changes does your code introduce? Put an `x` in all the boxes that apply:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds core functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Documentation (update in the documentation)

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [x] I have read the [CONTRIBUTION](https://omnisafe.readthedocs.io/en/latest/developer/contributing.html) guide. (**required**)
- [x] My change requires a change to the documentation.
- [x] I have updated the tests accordingly. (*required for a bug fix or a new feature*)
- [x] I have updated the documentation accordingly.
- [x] I have reformatted the code using `make format`. (**required**)
- [x] I have checked the code using `make lint`. (**required**)
- [x] I have ensured `make test` pass. (**required**)

### CODE REVIEW

**Review by zmsn-2077 — CHANGES_REQUESTED** [review]

File: omnisafe/algorithms/model_based/cap_pets.py [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+                    'Plan/episode_returns_min': info['Plan/episode_returns_min'],
+                    'Plan/episode_costs_max': info['Plan/episode_costs_max'],
+                    'Plan/episode_costs_mean': info['Plan/episode_costs_mean'],
+                    'Plan/episode_costs_min': info['Plan/episode_costs_min'],
+                    'Plan/var_penalty_max': info['Plan/var_penalty_max'],
+                    'Plan/var_penalty_mean': info['Plan/var_penalty_mean'],
+                    'Plan/var_penalty_min': info['Plan/var_penalty_min'],
+                },
+            )
+        assert action.shape == torch.Size(
```
Comment (rockmagma02) [inline omnisafe/algorithms/model_based/cap_pets.py:141]: consider use `self._logger.store(**info)` ?

File: omnisafe/algorithms/model_based/cce_pets.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+                    'Plan/feasible_num': info['Plan/feasible_num'],
+                    'Plan/episode_returns_max': info['Plan/episode_returns_max'],
+                    'Plan/episode_returns_mean': info['Plan/episode_returns_mean'],
+                    'Plan/episode_returns_min': info['Plan/episode_returns_min'],
+                    'Plan/episode_costs_max': info['Plan/episode_costs_max'],
+                    'Plan/episode_costs_mean': info['Plan/episode_costs_mean'],
+                    'Plan/episode_costs_min': info['Plan/episode_costs_min'],
+                },
+            )
+        assert action.shape == torch.Size(
```
Comment (rockmagma02) [inline omnisafe/algorithms/model_based/cce_pets.py:122]: consider use `self._logger.store(**info)`

File: omnisafe/algorithms/model_based/rce_pets.py [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+                    'Plan/feasible_num': info['Plan/feasible_num'],
+                    'Plan/episode_returns_max': info['Plan/episode_returns_max'],
+                    'Plan/episode_returns_mean': info['Plan/episode_returns_mean'],
+                    'Plan/episode_returns_min': info['Plan/episode_returns_min'],
+                    'Plan/episode_costs_max': info['Plan/episode_costs_max'],
+                    'Plan/episode_costs_mean': info['Plan/episode_costs_mean'],
+                    'Plan/episode_costs_min': info['Plan/episode_costs_min'],
+                },
+            )
+        assert action.shape == torch.Size(
```
Comment (rockmagma02) [inline omnisafe/algorithms/model_based/rce_pets.py:122]: consider use `self._logger.store(**info)`

File: omnisafe/algorithms/model_based/safeloop.py [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+                    'Plan/feasible_num': info['Plan/feasible_num'],
+                    'Plan/episode_returns_max': info['Plan/episode_returns_max'],
+                    'Plan/episode_returns_mean': info['Plan/episode_returns_mean'],
+                    'Plan/episode_returns_min': info['Plan/episode_returns_min'],
+                    'Plan/episode_costs_max': info['Plan/episode_costs_max'],
+                    'Plan/episode_costs_mean': info['Plan/episode_costs_mean'],
+                    'Plan/episode_costs_min': info['Plan/episode_costs_min'],
+                },
+            )
+        assert action.shape == torch.Size(
```
Comment (rockmagma02) [inline omnisafe/algorithms/model_based/safeloop.py:131]: consider use `self._logger.stroe(**info)`

File: omnisafe/algorithms/model_based/base/pets.py [hunk 5]
```diff
@@ -0,0 +1,458 @@
...
+        current_step,  # pylint: disable=unused-argument
+    ):
+        pass
+
+    def _update_policy(
+        self,
+        current_step,  # pylint: disable=unused-argument
+    ):
+        pass
+
```
Comment (rockmagma02) [inline omnisafe/algorithms/model_based/base/pets.py:227]: why pass here?
```suggestion
    def _algo_reset(
        self,
        current_step,  # pylint: disable=unused-argument
    ):
        pass

    def _update_policy(
        self,
        current_step,  # pylint: disable=unused-argument
    ):
        pass

```

File: omnisafe/algorithms/model_based/planner/arc.py [hunk 6]
```diff
@@ -0,0 +1,336 @@
...
+        )
+        return (
+            traj['actions']
+            .reshape(self._horizon, 1, *self._action_shape)
+            .clone()
+            .repeat([1, self._actor_traj, 1])
+        )
+
+    @torch.no_grad()
+    def _state_action_repeat(self, state: torch.Tensor, action: torch.Tensor) -> torch.Tensor:
```
Comment (zmsn-2077) [inline omnisafe/algorithms/model_based/planner/arc.py:146]: return list

File: omnisafe/algorithms/model_based/planner/cem.py [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+        ), 'Input returns dimension should be equal to (self._num_samples, 1)'
+        elite_idxs = torch.topk(mean_episode_returns.squeeze(1), self._num_elites, dim=0).indices
+        elite_values, elite_actions = mean_episode_returns[elite_idxs], actions[:, elite_idxs]
+
+        info = {
+            'Plan/episode_returns_max': mean_episode_returns.max().item(),
+            'Plan/episode_returns_mean': mean_episode_returns.mean().item(),
+            'Plan/episode_returns_min': mean_episode_returns.min().item(),
+        }
+        # info['elite_idxs'] = elite_idxs
```
Comment (zmsn-2077) [inline omnisafe/algorithms/model_based/planner/cem.py:178]: del unuseful comments

**Review by zmsn-2077 — APPROVED** [review]
LGTM.

File: benchmarks/model-based/README.md [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+```bash
+cd examples/benchmarks
+python run_experiment_grid.py
+```
+
+You can set the path of ``examples/benchmarks/experimrnt_grid.py`` :
+example:
+
+```python
+path ='/home/username/omnisafe/omnisafe/examples/benchmarks/exp-x/Model-Based-Benchmarks'
```
Comment (Gaiejj) [inline benchmarks/model-based/README.md:69]: May be relative path better. e.g. `./examples/benchmarks/exp-x/Model-Based-Benchmarks`

**Review by rockmagma02 — APPROVED** [review]

File: omnisafe/adapter/modelbased_adapter.py [hunk 9]
```diff
@@ -0,0 +1,382 @@
...
+        self._ep_ret: torch.Tensor
+        self._ep_cost: torch.Tensor
+        self._ep_len: torch.Tensor
+        self._current_obs, _ = self.reset()
+        self._max_ep_len = 1000
+        self._reset_log()
+        self._last_dynamics_update = 0
+        self._last_policy_update = 0
+        self._last_eval = 0
+        self._first_log = False
```
Comment (Gaiejj) [inline omnisafe/adapter/modelbased_adapter.py:118]:
```suggestion
        self._first_log: bool = False
```
The same with others.

File: omnisafe/adapter/modelbased_adapter.py [hunk 10]
```diff
@@ -0,0 +1,382 @@
...
+            self._env.get_cost_from_obs_tensor(obs)
+            if hasattr(self._env, 'get_cost_from_obs_tensor')
+            else None
+        )
+
+    def get_lidar_from_coordinate(self, obs: torch.Tensor) -> torch.Tensor | None:
+        """Get lidar from numpy coordinate.
+
+        Args:
+            obs (np.ndarray): The observation.
```
Comment (Gaiejj) [inline omnisafe/adapter/modelbased_adapter.py:148]:
```suggestion
            obs (torch.Tensor): The observation.
```

File: omnisafe/adapter/modelbased_adapter.py [hunk 11]
```diff
@@ -0,0 +1,382 @@
...
+        """
+        return (
+            self._env.get_lidar_from_coordinate(obs)
+            if hasattr(self._env, 'get_lidar_from_coordinate')
+            else None
+        )
+
+    def render(self, *args: str, **kwargs: int) -> Any:
```
Comment (Gaiejj) [inline omnisafe/adapter/modelbased_adapter.py:156]: why the ``**kwargs`` is ``int`` type ?

File: omnisafe/adapter/modelbased_adapter.py [hunk 12]
```diff
@@ -0,0 +1,382 @@
...
+        self._first_log = True
+        logger.store(
+            **{
+                'Metrics/EpRet': self._ep_ret,
+                'Metrics/EpCost': self._ep_cost,
+                'Metrics/EpLen': self._ep_len,
+            },
+        )
+
+    def _reset_log(self, idx: int | None = None) -> None:  # pylint: disable=unused-argument
```
Comment (Gaiejj) [inline omnisafe/adapter/modelbased_adapter.py:374]: why unused-argument ``idx`` maintained ?

**Review by Gaiejj — APPROVED** [review]
Some problems still exist, but it doesn't matter, I will open another PR to fix my issue and I will request your review.

**Comment by zmsn-2077** [comment]
Open a fix request, then fix this code details.

**Pull request merged**